drivers/spi/spi_npcx_spip.c

```diff
@@ -0,0 +1,436 @@
+		data->bytes_per_frame = 2;
+	} else {
+		LOG_ERR("Only support word sizes either 8 or 16 bits");
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_SPI_EXTENDED_MODES) &&
+	    (operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
+		LOG_ERR("Only single line mode is supported");
+		return -ENOTSUP;
+	}
+
+	/* Set the endianness */
+	if (operation & SPI_TRANSFER_LSB) {
+		LOG_ERR("Shift out with LSB is not supported");
+		return -ENOTSUP;
+	}
+
+	/*
+	 * Set CPOL and CPHA.
+	 * The following is how to map npcx spip control register to CPOL and CPHA
+	 *   CPOL    CPHA  |  SCIDL    SCM
+	 *   -----------------------------
+	 *    0       0    |    0       0
+	 *    0       1    |    0       1
+	 *    1       0    |    1       1
+	 *    1       1    |    1       0
+	 */
+	if (operation & SPI_MODE_CPOL) {
+		reg_base->SPIP_CTL1 |= BIT(NPCX_SPIP_CTL1_SCIDL);
+	} else {
+		reg_base->SPIP_CTL1 &= ~BIT(NPCX_SPIP_CTL1_SCIDL);
+	}
+	if ((operation & SPI_MODE_CPOL) ^ (operation & SPI_MODE_CPHA)) {
+		reg_base->SPIP_CTL1 |= BIT(NPCX_SPIP_CTL1_SCM);
+	} else {
+		reg_base->SPIP_CTL1 &= ~BIT(NPCX_SPIP_CTL1_SCM);
+	}
+
+	/* Set the SPI frequency */
+	prescaler_dividor = data->src_clock_freq / 2 / spi_cfg->frequency;
+	if (prescaler_dividor >= 1) {
+		prescaler_dividor -= 1;
+	}
+	if (prescaler_dividor >= 0x7F) {
+		LOG_INF("SPI divider exceeds the max allowed value.");
+		return -ENOTSUP;
+	}
+	SET_FIELD(reg_base->SPIP_CTL1, NPCX_SPIP_CTL1_SCDV, prescaler_dividor);
+
+	data->ctx.config = spi_cfg;
+
+	return 0;
+}
+
+static void spi_npcx_spip_process_tx_buf(struct spi_npcx_spip_data *const data, uint16_t *tx_frame)
+{
+	/* Get the tx_frame from tx_buf only when tx_buf != NULL */
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		if (data->bytes_per_frame == 1) {
+			*tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+		} else {
+			*tx_frame = UNALIGNED_GET((uint16_t *)(data->ctx.tx_buf));
+		}
+	}
+	/*
+	 * The update is ignored if TX is off (tx_len == 0).
+	 * Note: if tx_buf == NULL && tx_len != 0, the update still counts.
+	 */
+	spi_context_update_tx(&data->ctx, data->bytes_per_frame, 1);
+}
+
+static void spi_npcx_spip_process_rx_buf(struct spi_npcx_spip_data *const data, uint16_t rx_frame)
+{
+	if (spi_context_rx_buf_on(&data->ctx)) {
+		if (data->bytes_per_frame == 1) {
+			UNALIGNED_PUT(rx_frame, (uint8_t *)data->ctx.rx_buf);
+		} else {
+			UNALIGNED_PUT(rx_frame, (uint16_t *)data->ctx.rx_buf);
+		}
+	}
+	spi_context_update_rx(&data->ctx, data->bytes_per_frame, 1);
+}
+
+#ifndef CONFIG_SPI_NPCX_SPIP_INTERRUPT
+static void spi_npcx_spip_xfer_frame(const struct device *dev)
+{
+	const struct spi_npcx_spip_cfg *const config = dev->config;
+	struct spip_reg *const reg_base = config->reg_base;
+	struct spi_npcx_spip_data *const data = dev->data;
+	uint16_t tx_frame = SPI_NPCX_SPIP_TX_NOP;
+	uint16_t rx_frame;
+
+	spi_npcx_spip_process_tx_buf(data, &tx_frame);
+
+	while (IS_BIT_SET(reg_base->SPIP_STAT, NPCX_SPIP_STAT_BSY)) {
+		/* NOP */
+	}
```

Comment:
no timeout? you could use `WAIT_FOR`, put some reasonable timeout, return an error and handle it in the caller